Updating make scripts to be "DRY" by moving "WET" code to common files

wpatterson-witekio opened the pull request.

Commit:
Updating make scripts to be "DRY" by moving "WET" code to common files

narrucmot commented:
Why delete the "rebuild" scripts?  We need to keep those.  They get copied into the project folder and this is how users rebuild the BSP once they start making their own customizations.

wpatterson-witekio commented:
Neither the rebuild feature nor any naming has changed, so to a user they can run ./make_foo.sh and later from their project folder run ./rebuild_foo.sh as before. 

TLDR;) If you want we could have a quick call, it's really quite simple since there was so much code duplication, but here's the long winded changes.

It helps to pull up the scripts in a diff tool to see how much is duplicated, which also makes this refactor very easy.

rebuild_common.sh is just common.sh without all the project setup functions.
I renamed rebuild_common.sh to just build.sh and moved it up. I removed all duplicated functions from common.sh and instead common.sh sources build.sh.

After all environment setup(sourcing, settings vars, etc) the main entry point/build function is `make_board`(this is true of `make_foo.sh` and `rebuild_foo.sh` I'll explain how this works)
All the rebuild scripts are the same as the make scripts, except the don't call the project setup functions or package bsp

For `make_foo.sh`, 
```
make_board()
{
    setup_project
    build_images
    package_bsp
}
```

For `rebuild_foo.sh`, 
```
make_board()
{
    verify_environment
    build_images
}
```
Otherwise, everything is exactly the same.

If you look at a new `make_foo.sh`, you'll see it's really just 3 lines

```
source ${MAIN_SCRIPT_FOLDER}/conf/minized.sh
source ${MAIN_SCRIPT_FOLDER}/common.sh
make_board
```

To support rebuilding, `package_bsp` does:
```
   mkdir conf
    cp ${PETALINUX_SCRIPTS_FOLDER}/conf/${PETALINUX_BOARD_FAMILY}.sh conf/
    cp ${PETALINUX_SCRIPTS_FOLDER}/build.sh ./common.sh
    cp ${PETALINUX_SCRIPTS_FOLDER}/make_${HDL_BOARD_NAME}_${HDL_PROJECT_NAME}.sh rebuild_${HDL_BOARD_NAME}_${HDL_PROJECT_NAME}.sh
```
 This replicates the initial scripts structure by creating a conf directory containing the family conf, `make_foo.sh` is copied over as `rebuild_foo.sh` and `build.sh` is copied over as `common.sh`(It has all the common build functions and project setup is no longer used)

Now `./rebuild_foo.sh` does the rebuild using it's definition of `make_board` which does not setup up the project or package the bsp, which matches the previous rebuild behavior.

tnizan-witekio commented:
I like the idea of making the make scripts smaller. For example, I like the 'setup_project' function.

However, I think some variables needs to stay in the 'board' make script, so that it's easier for users to see what will be built, and to be able to quickly modify those.

For example:
 
PETALINUX_PROJECT_ROOT_NAME: I like to be able to modify this variable to create a new project with a different name 

PETALINUX_BUILD_IMAGE=avnet-image-full: I prefer to see it in each script, so that the user knows which image will be built. The user can easily add a new image for a specific board and change this value.

same for KEEP_CACHE, KEEP_WORK, DEBUG,  NO_BIT_OPTION, BOOT_METHOD, BOOT_SUFFIX, INITRAMFS_IMAGE.
